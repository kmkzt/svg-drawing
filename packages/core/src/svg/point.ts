import type { VectorObject, PointClass, PointObject } from '../types'

/**
 * Maximum number of decimal places is 5
 *
 * @todo Changeable parameter
 */
const coeffecient = 100000
const roundNumber = (num: number) => Math.round(num * coeffecient) / coeffecient

export class Point implements PointClass {
  private _x: number
  private _y: number
  constructor(x: number, y: number) {
    this._x = roundNumber(x)
    this._y = roundNumber(y)
  }

  public get x() {
    return this._x
  }

  public get y() {
    return this._y
  }

  public scale(r: number) {
    return new Point(this._x * r, this._y * r)
  }

  public scaleX(r: number) {
    return new Point(this._x * r, this._y)
  }

  public scaleY(r: number) {
    return new Point(this._x, this._y * r)
  }

  public add(p: PointObject) {
    return new Point(this._x + p.x, this._y + p.y)
  }

  public sub(p: PointObject) {
    return new Point(this._x - p.x, this._y - p.y)
  }

  public eql(p: PointObject) {
    return this._x === p.x && this._y === p.y
  }

  public clone() {
    return new Point(this._x, this._y)
  }

  public toJson() {
    return {
      x: this._x,
      y: this._y,
    }
  }

  /**
   * ### Calculate coordinate from vector.
   *
   * ```ts
   * const point = Point.fromVector({ value: 1, 0 })
   *
   * console.log(point.x) // 0
   * console.log(point.y) // 1
   * ```
   */
  public static fromVector({ value, angle }: VectorObject): Point {
    const theta = (angle * Math.PI) / 180
    return new Point(Math.cos(theta) * value, Math.sin(theta) * value)
  }
}
