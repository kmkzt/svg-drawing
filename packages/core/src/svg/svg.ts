import type { SvgOption, PathClass, SvgClass } from '../types'

export class Svg implements SvgClass {
  public paths: PathClass[]
  public width: number
  public height: number
  public background?: string

  constructor({ width, height, background }: SvgOption) {
    this.paths = []
    this.width = width
    this.height = height
    this.background = background
  }

  /** @todo Implement to be resized according to height */
  public resize({ width, height }: { width: number; height: number }) {
    this.scale(width / this.width)
    this.width = width
    this.height = height
  }

  private scale(r: number) {
    if (r !== 1) {
      for (let i = 0; i < this.paths.length; i += 1) {
        this.paths[i].scale(r)
      }
    }
    return this
  }

  public addPath(pa: PathClass | PathClass[]) {
    if (Array.isArray(pa)) {
      this.paths.push(...pa)
    } else {
      this.paths.push(pa)
    }
    return this
  }

  public getPath(key: string): PathClass | undefined {
    return this.paths.find((p) => p.key === key)
  }

  public updatePath(path: PathClass) {
    const index = this.paths.findIndex((p) => p.key === path.key)
    if (index !== -1) {
      this.paths[index] = path
    }
    return this
  }

  public deletePath(deletePath: PathClass) {
    this.paths = this.paths.filter((path) => path.key !== deletePath.key)
    return this
  }

  public clonePaths() {
    return this.paths.map((p) => p.clone())
  }

  public toJson() {
    return {
      width: this.width,
      height: this.height,
      background: this.background,
      paths: this.paths.map((p) => p.toJson()),
    }
  }

  /**
   * Copy resized paths.
   *
   * @example Const drawSvg = new Svg() const animateSvg = new Svg().copy(drawSvg)
   */
  public copy(svg: SvgClass) {
    this.paths = svg.clonePaths()
    if (svg.width && this.width) {
      this.scale(this.width / svg.width)
    }
    return this
  }

  public clone() {
    const svg = new Svg({
      width: this.width,
      height: this.height,
      background: this.background,
    })
    svg.addPath(this.clonePaths())
    return svg
  }
}
