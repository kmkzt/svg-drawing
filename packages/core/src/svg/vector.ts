import type { VectorClass, PointObject } from '../types'

export class Vector implements VectorClass {
  constructor(public readonly value: number, public readonly angle: number) {}

  scale(r: number) {
    return new Vector(this.value * r, this.angle)
  }

  rotate(a: number) {
    return new Vector(this.value, this.angle + a)
  }

  toJson() {
    return {
      value: this.value,
      angle: this.angle,
    }
  }

  /**
   * ### Calculate angle and absolute value from origin of coordinates
   *
   * ```ts
   * const vector = Vector.fromPoint({ x: 1, y: 1 })
   *
   * console.log(vector.angle) // 45
   * console.log(vector.value) // 1.4142135623730951
   * ```
   */
  public static fromPoint({ x, y }: PointObject): Vector {
    return new Vector(
      Math.sqrt(Math.pow(x, 2.0) + Math.pow(y, 2.0)),
      Math.atan2(y, x) * (180 / Math.PI)
    )
  }
}
