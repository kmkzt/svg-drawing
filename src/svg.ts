import { roundUp } from './utils/roundUp'
import { camel2kebab } from './utils/camel2kebab'
import { kebab2camel } from './utils/kebab2camel'
import { svg2base64 } from './utils/svg2base64'
import {
  createSvgElement,
  createSvgChildElement
} from './utils/createSvgElement'

const isNaN = (num: number) => num !== num

export class Point {
  public x: number
  public y: number
  constructor(x: number, y: number) {
    this.x = x
    this.y = y
  }

  public toVector(): Vector {
    const value = Math.sqrt(Math.pow(this.x, 2.0) + Math.pow(this.y, 2.0))
    const angle = Math.atan2(this.y, this.x)
    return new Vector(value, angle)
  }

  public scale(r: number): Point {
    return new Point(this.x * r, this.y * r)
  }

  public add(p: Point): Point {
    return new Point(this.x + p.x, this.y + p.y)
  }

  public sub(p: Point): Point {
    return new Point(this.x - p.x, this.y - p.y)
  }

  public eql(p: Point): boolean {
    return this.x === p.x && this.y === p.y
  }

  public clone(): Point {
    return new Point(this.x, this.y)
  }
}

export const COMMAND_TYPE = {
  MOVE: 'M', // M 0 0
  MOVE_RELATIVE: 'm', // m 0 0
  LINE: 'L', // L 1 1
  LINE_RELATIVE: 'l', // l 1 1
  CURVE: 'C', // C 1 1 2 2 3 3
  CURVE_RELATIVE: 'c', // c 1 1 2 2 3 3
  CLOSE: 'Z', // Z, z
  HORIZONTAL: 'H', // H 10
  HORIZONTAL_RELATIVE: 'h', // h 10
  VERTICAL: 'V', // V 20
  VERTICAL_RELATIVE: 'v', // v 20
  ARC_CURVE: 'A', // A 6 4 10 0 1 14 10
  ARC_CURVE_RELATIVE: 'a', // A 6 4 10 0 1 14 10
  QUADRATIC_CURVE: 'Q', // Q 10 60 10 30
  QUADRATIC_CURVE_RELATIVE: 'q' // q 10 60 10 30
} as const

type COMMAND = typeof COMMAND_TYPE[keyof typeof COMMAND_TYPE]
// TODO: compatible COMMAND_TYPE
export class Command {
  public type: COMMAND
  public value: number[]
  // TODO: Convert data format to number array.
  constructor(type: COMMAND, value: number[] = []) {
    this.value = value
    this.type = type
  }

  public set cr(po: Point | undefined) {
    if (!po) return
    if ((this.type !== 'C' && this.type !== 'c') || this.value.length !== 6) {
      return
    }
    this.value.splice(2, 1, po.x)
    this.value.splice(3, 1, po.y)
  }
  public get cr(): Point | undefined {
    if ((this.type !== 'C' && this.type !== 'c') || this.value.length !== 6) {
      return undefined
    }
    const [x, y] = this.value.slice(2, 4)
    return new Point(x, y)
  }
  public set cl(po: Point | undefined) {
    if (!po) return
    if ((this.type !== 'C' && this.type !== 'c') || this.value.length !== 6) {
      return
    }
    this.value.splice(0, 1, po.x)
    this.value.splice(1, 1, po.y)
  }
  public get cl(): Point | undefined {
    if ((this.type !== 'C' && this.type !== 'c') || this.value.length !== 6) {
      return undefined
    }
    const [x, y] = this.value.slice(0, 2)
    return new Point(x, y)
  }

  public set point(po: Point | undefined) {
    if (!po) return
    this.value.splice(this.value.length - 2, 1, po.x)
    this.value.splice(this.value.length - 1, 1, po.y)
  }
  public get point(): Point | undefined {
    const xy = this.value.slice(this.value.length - 2)
    return xy.length === 2 ? new Point(xy[0], xy[1]) : undefined
  }

  public toString(): string {
    return `${this.type} ${this.value.map(v => roundUp(v)).join(' ')}`
  }

  public scale(r: number): Command {
    const upd = new Command(
      this.type,
      this.value.map(p => p * r)
    )
    return upd
  }

  public clone(): Command {
    const copy = new Command(this.type, this.value.slice())
    return copy
  }
}

export class Vector {
  public angle: number
  public value: number
  constructor(v: number, a: number) {
    this.value = v
    this.angle = a
  }

  public toPoint(): Point {
    const x = Math.cos(this.angle) * this.value
    const y = Math.sin(this.angle) * this.value
    return new Point(x, y)
  }

  public scale(r: number): Vector {
    return new Vector(this.value * r, this.angle)
  }
}

export interface PathOption {
  strokeWidth?: number
  attrs?: Attrs
  stroke?: string
  fill?: string
}

interface Attrs {
  [key: string]: string
}
export interface PathObject {
  attrs: Attrs
  d: string
}

const defaultAttrs: Attrs = {
  strokeLinecap: 'round', // 'mitter'
  strokeLinejoin: 'round' // 'square'
}
/**
 * TODO: refactor command.
 * The following commands are not supported. Cannot support commands that use `M` or` z` more than once
 * `M 0 0 L 1 1 Z M 1 1 L 2 2 Z`
 */
export class Path {
  public strokeWidth: number
  public attrs: { [key: string]: string }
  public stroke: string
  public fill: string
  public commands: Command[]

  constructor({ strokeWidth, fill, stroke, attrs }: PathOption = {}) {
    this.strokeWidth = strokeWidth ?? 1
    this.stroke = stroke ?? '#000'
    this.fill = fill ?? 'none'
    this.attrs = {
      ...defaultAttrs,
      ...attrs
    }
    this.commands = []
  }

  public scale(r: number): this {
    this.commands = this.commands.map((c: Command) => c.scale(r))
    this.strokeWidth *= r
    return this
  }

  public addCommand(param: Command): this {
    this.commands.push(param)
    return this
  }
  public getCommandString(): string {
    if (this.commands.length === 0) return ''
    return this.commands
      .map((com: Command, _i: number) => com.toString())
      .join(' ')
      .trim()
  }

  // TODO: Valid Command type
  public parseCommandString(d: string): void {
    this.commands = []
    let type: COMMAND | null = null
    let value: number[] = []
    const c = d.split(' ')
    const checkType = (c: any): COMMAND | null =>
      Object.values(COMMAND_TYPE).includes(c) ? c : null
    for (let i = 0; i < c.length; i += 1) {
      const t = checkType(c[i])
      // COMMAND Parse
      if (t) {
        if (!type) {
          type = t
          continue
        }
        this.commands.push(new Command(type, value))
        type = t
        value = []
        continue
      }
      if (isNaN(+c[i])) {
        return
      }
      value.push(+c[i])
    }

    if (type !== null) {
      this.commands.push(new Command(type, value))
    }
  }

  public parsePathElement(pEl: SVGPathElement): this {
    this.strokeWidth = 1
    this.stroke = '#000'
    this.fill = 'none'
    this.attrs = defaultAttrs
    for (let i = 0; i < pEl.attributes.length; i += 1) {
      const attr: Attr | null = pEl.attributes.item(i)
      if (!attr) continue
      if (attr.name === 'd') {
        this.parseCommandString(attr.value)
        continue
      }
      if (attr.name === 'stroke-width') {
        this.strokeWidth = Number(attr.value)
        continue
      }
      if (attr.name === 'fill') {
        this.fill = attr.value
        continue
      }
      if (attr.name === 'stroke') {
        this.stroke = attr.value
        continue
      }
      this.attrs = {
        ...this.attrs,
        [kebab2camel(attr.name)]: attr.value
      }
    }
    return this
  }

  public toJson(): PathObject {
    return {
      attrs: this.attrs,
      d: this.getCommandString()
    }
  }
  public toElement(): SVGElement {
    return createSvgChildElement('path', {
      fill: this.fill,
      stroke: this.stroke,
      ['stroke-width']: String(this.strokeWidth),
      d: this.getCommandString(),
      ...Object.entries(this.attrs).reduce(
        (acc, [key, val], _i) => ({
          ...acc,
          [camel2kebab(key)]: val
        }),
        {}
      )
    })
  }

  public clone(): Path {
    const path = new Path({
      strokeWidth: this.strokeWidth,
      fill: this.fill,
      stroke: this.stroke,
      attrs: { ...this.attrs }
    })
    this.commands.map(c => {
      path.commands.push(c.clone())
    })
    return path
  }
}

export interface SvgOption {
  width: number
  height: number
}

export interface SvgObject {
  width: number
  height: number
  paths: PathObject[]
}
export class Svg {
  public paths: Path[]
  public width: number
  public height: number
  constructor({ width, height }: SvgOption) {
    this.paths = []
    this.width = width
    this.height = height
  }

  public scalePath(r: number): this {
    if (r !== 1) {
      for (let i = 0; i < this.paths.length; i += 1) {
        this.paths[i].scale(r)
      }
    }
    return this
  }

  public addPath(pa: Path): this {
    this.paths.push(pa)
    return this
  }
  public clonePaths(): Path[] {
    return this.paths.map(p => p.clone())
  }

  public updatePath(pa: Path, i?: number): this {
    const updateIndex = i || this.paths.length - 1
    if (updateIndex < 0) this.paths.push(pa)
    this.paths[updateIndex] = pa
    return this
  }

  public toJson(): SvgObject {
    return {
      width: this.width,
      height: this.height,
      paths: this.paths.map(p => p.toJson())
    }
  }

  public copy(svg: any extends Svg ? Svg : never): this {
    this.paths = svg.clonePaths()
    if (svg.width && this.width) {
      this.scalePath(svg.width / this.width)
    }
    return this
  }

  public toElement(): SVGSVGElement {
    return createSvgElement(
      { width: String(this.width), height: String(this.height) },
      this.paths.map(p => p.toElement())
    )
  }
  public toBase64(): string {
    return svg2base64(this.toElement().outerHTML)
  }

  public parseSVGString(svgStr: string): this {
    const svgEl: SVGSVGElement | null = new DOMParser()
      .parseFromString(svgStr, 'image/svg+xml')
      .querySelector('svg')
    if (!svgEl) {
      this.paths = []
      return this
    }
    return this.parseSVGElement(svgEl)
  }

  public parseSVGElement(svgEl: SVGSVGElement): this {
    const update: Path[] = []
    svgEl.querySelectorAll('path').forEach(pEl => {
      const pa = new Path().parsePathElement(pEl)
      if (pa.commands.length !== 0) {
        update.push(pa)
      }
    })
    this.paths = update
    const width = Number(svgEl.getAttribute('width'))
    if (width && this.width) {
      this.scalePath(this.width / width)
    }
    return this
  }
}
