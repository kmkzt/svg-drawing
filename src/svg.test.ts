import { Point, Vector, Path, Svg, Command, CommandType } from './svg'

describe('svg.ts', () => {
  describe('Point', () => {
    it('add', () => {
      const po = new Point(1.0, 1.0).add(new Point(2.0, 2.0))
      expect(po.x).toBe(3.0)
      expect(po.y).toBe(3.0)
    })

    it('sub', () => {
      const po = new Point(3.0, 3.0).sub(new Point(1.0, 1.0))
      expect(po.x).toBe(2.0)
      expect(po.y).toBe(2.0)
    })

    it('scale', () => {
      const po = new Point(1.0, 1.0).scale(3)
      expect(po.x).toBe(3.0)
      expect(po.y).toBe(3.0)
    })

    it('toVector', () => {
      const vec = new Point(1.0, 1.0).toVector()
      expect(vec.value).toBe(1.4142135623730951)
      expect(vec.angle).toBe(0.7853981633974483)
    })

    it('clone', () => {
      const po = new Point(0, 0)
      const clone = po.clone()
      clone.x = 2
      expect(po.x).toBe(0)
      expect(clone.x).toBe(2)
    })
  })
  describe('Vector', () => {
    it('toPoint', () => {
      const po = new Vector(1.4142135, 0.7853982).toPoint()
      expect(po.x).toBe(1.0)
      expect(po.y).toBe(1.0)
    })
    it('scale', () => {
      const vec = new Vector(1.0, 0.5).scale(0.3)
      expect(vec.value).toBe(0.3)
      expect(vec.angle).toBe(0.5)
    })
  })
  describe('Command', () => {
    it('MOVE', () => {
      const cmd = new Command(CommandType.MOVE, new Point(0, 0))
      expect(cmd.toString()).toBe('M 0 0')
    })
    it('LINE', () => {
      const cmd = new Command(CommandType.LINE, new Point(1, 1))
      expect(cmd.toString()).toBe('L 1 1')
    })
    it('CURVE', () => {
      const cmd = new Command(CommandType.CURVE, new Point(1, 1))
      expect(cmd.toString()).toBe('L 1 1')
      cmd.cl = new Point(0.25, 0.25)
      cmd.cr = new Point(0.75, 0.25)
      expect(cmd.toString()).toBe('C 0.25 0.25 0.75 0.25 1 1')
    })

    describe('clone', () => {
      let cmd
      let clone
      beforeEach(() => {
        cmd = new Command(CommandType.CURVE, new Point(1, 1))
        cmd.cl = new Point(0.25, 0.25)
        cmd.cr = new Point(0.75, 0.25)
        clone = cmd.clone()
      })
      it('point is not overwritten', () => {
        clone.x = 1.5
        expect(cmd.point.x).toBe(1)
        expect(clone.point.x).toBe(1)
      })
      it('cl is not overwritten', () => {
        clone.cl = new Point(0.1, 0.1)
        expect(cmd.cl.x).toBe(0.25)
        expect(clone.cl.x).toBe(0.1)
      })
      it('cr is not overwritten', () => {
        clone.cr = new Point(1.2, 0.1)
        expect(cmd.cr.x).toBe(0.75)
        expect(clone.cr.x).toBe(1.2)
      })
    })
  })
  describe('Path', () => {
    let path: Path
    beforeEach(() => {
      path = new Path({ strokeWidth: 1 })
        .addCommand(new Point(1, 1))
        .addCommand(new Point(2, 2))
    })
    it('addCommand', () => {
      expect(path.commands.length).toBe(2)
      expect(path.commands[0].type).toBe(CommandType.MOVE)
      expect(path.commands[0].point.x).toBe(1)
      expect(path.commands[0].point.y).toBe(1)
    })
    it('scale', () => {
      path.scale(2)
      expect(path.strokeWidth).toBe(2)
      expect(path.commands[0].type).toBe(CommandType.MOVE)
      expect(path.commands[0].point.x).toBe(2)
      expect(path.commands[0].point.y).toBe(2)
    })
    describe('parsePathElement', () => {
      it('success pattern', () => {
        const pathEl = document.createElementNS(
          'http://www.w3.org/2000/svg',
          'path'
        )
        pathEl.setAttribute('fill', '#00f')
        pathEl.setAttribute('stroke', '#f00')
        pathEl.setAttribute('stroke-width', '4')
        pathEl.setAttribute('d', 'M 0 0 L 1 1 C 2 2 2 4 6 0')
      })
    })
    describe('parseCommandString', () => {
      beforeEach(() => {
        path = new Path()
      })
      it('success pattern', () => {
        path.parseCommandString('M 0 0 L 1 1 C 2 2 2 4 6 0')
        expect(path.commands).toMatchSnapshot()
        expect(path.close).toBe(false)
      })
      it('close command', () => {
        path.parseCommandString('M 0 0 L 1 1 C 2 2 2 4 6 0 Z')
        expect(path.commands).toMatchSnapshot()
        expect(path.close).toBe(true)
      })
      it('failed pattern', () => {
        path.parseCommandString('M a b')
        expect(path.commands).toMatchSnapshot()
        path.parseCommandString('M 0 0 C 0 1 2')
        expect(path.commands).toMatchSnapshot()
      })
    })
    it('clone', () => {
      const origin = new Path({ strokeWidth: 1 }).addCommand(new Point(1, 1))
      const clone = origin.clone().addCommand(new Point(2, 2))
      clone.commands[0].point.x = 3
      expect(origin.commands.length).toBe(1)
      expect(clone.commands.length).toBe(2)
      expect(origin.commands[0].point.x).toBe(1)
      expect(clone.commands[0].point.x).toBe(3)
    })
    describe('toJson and toElement', () => {
      const path = new Path({ curve: true, close: false })
        .addCommand(new Point(0, 0))
        .addCommand(new Point(1, 1))
        .addCommand(new Point(2, 1))
        .addCommand(new Point(3, 0))

      it('toJson', () => {
        expect(path.toJson()).toMatchSnapshot()
      })
      it('toElement', () => {
        expect(path.toElement()).toMatchSnapshot()
      })
    })
    describe('commands parameter and getCommandString', () => {
      describe('Line', () => {
        const path = new Path({ curve: false })
          .addCommand(new Point(0, 0))
          .addCommand(new Point(1, 1))
          .addCommand(new Point(-1, -1))
        it('Normal', () => {
          path.close = false
          expect(path.commands).toMatchSnapshot()
          expect(path.getCommandString()).toMatchSnapshot()
        })
        it('Close', () => {
          path.close = true
          expect(path.commands).toMatchSnapshot()
          expect(path.getCommandString()).toMatchSnapshot()
        })
      })
      describe('curve', () => {
        const path = new Path({ curve: true })
          .addCommand(new Point(0, 0))
          .addCommand(new Point(1, 1))
          .addCommand(new Point(2, 1))
          .addCommand(new Point(3, 0))
        it('Normal', () => {
          path.close = false
          expect(path.commands).toMatchSnapshot()
          expect(path.getCommandString()).toMatchSnapshot()
        })
        it('Close', () => {
          path.close = true
          expect(path.commands).toMatchSnapshot()
          expect(path.getCommandString()).toMatchSnapshot()
        })
      })
    })
  })
  describe('Svg', () => {
    let svg: Svg
    beforeEach(() => {
      svg = new Svg({ width: 500, height: 500 })
        .addPath(
          new Path({ curve: true, close: false })
            .addCommand(new Point(0, 0))
            .addCommand(new Point(1, 1))
            .addCommand(new Point(2, 1))
            .addCommand(new Point(3, 0))
        )
        .addPath(
          new Path({ curve: false, close: true })
            .addCommand(new Point(4, 4))
            .addCommand(new Point(9, 4))
            .addCommand(new Point(9, 8))
            .addCommand(new Point(3, 0))
        )
    })

    it('parseSVGString', () => {
      expect(
        new Svg({ width: 400, height: 400 }).parseSVGString(`
      <svg width="200" height="200">
        <path fill="#f00" stroke="#00f" stroke-width="4" d="M 1 1 L 2 2 C 3 3 5 3 7 3 Z"></path>
      </svg>`)
      ).toMatchSnapshot()
    })
    it('parseSVGElement', () => {
      const svgEl = document.createElementNS(
        'http://www.w3.org/2000/svg',
        'svg'
      )
      svgEl.setAttribute('width', '200')
      svgEl.setAttribute('height', '200')
      const pathEl = document.createElementNS(
        'http://www.w3.org/2000/svg',
        'path'
      )
      pathEl.setAttribute('fill', '#f00')
      pathEl.setAttribute('stroke', '#00f')
      pathEl.setAttribute('stroke-width', '4')
      pathEl.setAttribute('d', 'M 1 1 L 2 2 C 3 3 5 3 7 3 Z')
      svgEl.appendChild(pathEl)
      expect(
        new Svg({ width: 400, height: 400 }).parseSVGElement(svgEl)
      ).toMatchSnapshot()
    })
    // TODO: Fix width, height
    it('toElement', () => {
      expect(svg.toElement()).toMatchSnapshot()
    })
    it('toJson', () => {
      expect(svg.toJson()).toMatchSnapshot()
    })
    // TODO: replace image snapshot
    it('toBase64', () => {
      expect(svg.toBase64()).toMatchSnapshot()
    })
  })
})
