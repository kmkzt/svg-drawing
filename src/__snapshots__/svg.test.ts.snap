// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`svg.ts Path commands parameter and getCommandString Line Close 1`] = `
Array [
  Command {
    "point": Point {
      "x": 0,
      "y": 0,
    },
    "type": "M",
  },
  Command {
    "point": Point {
      "x": 1,
      "y": 1,
    },
    "type": "L",
  },
  Command {
    "point": Point {
      "x": -1,
      "y": -1,
    },
    "type": "L",
  },
]
`;

exports[`svg.ts Path commands parameter and getCommandString Line Close 2`] = `"M 0 0 L 1 1 L -1 -1 Z"`;

exports[`svg.ts Path commands parameter and getCommandString Line Normal 1`] = `
Array [
  Command {
    "point": Point {
      "x": 0,
      "y": 0,
    },
    "type": "M",
  },
  Command {
    "point": Point {
      "x": 1,
      "y": 1,
    },
    "type": "L",
  },
  Command {
    "point": Point {
      "x": -1,
      "y": -1,
    },
    "type": "L",
  },
]
`;

exports[`svg.ts Path commands parameter and getCommandString Line Normal 2`] = `"M 0 0 L 1 1 L -1 -1"`;

exports[`svg.ts Path commands parameter and getCommandString curve Close 1`] = `
Array [
  Command {
    "point": Point {
      "x": 0,
      "y": 0,
    },
    "type": "M",
  },
  Command {
    "cl": Point {
      "x": 0.2,
      "y": 0.2,
    },
    "cr": Point {
      "x": 0.6,
      "y": 0.8,
    },
    "point": Point {
      "x": 1,
      "y": 1,
    },
    "type": "C",
  },
  Command {
    "cl": Point {
      "x": 1.4,
      "y": 1.2,
    },
    "cr": Point {
      "x": 1.6,
      "y": 1.2,
    },
    "point": Point {
      "x": 2,
      "y": 1,
    },
    "type": "C",
  },
  Command {
    "cl": Point {
      "x": 2.4,
      "y": 0.8,
    },
    "cr": Point {
      "x": 2.8,
      "y": 0.2,
    },
    "point": Point {
      "x": 3,
      "y": 0,
    },
    "type": "C",
  },
]
`;

exports[`svg.ts Path commands parameter and getCommandString curve Close 2`] = `"M 0 0 C 0.2 0.2 0.6 0.8 1 1 C 1.4 1.2 1.6 1.2 2 1 C 2.4 0.8 2.8 0.2 3 0 Z"`;

exports[`svg.ts Path commands parameter and getCommandString curve Normal 1`] = `
Array [
  Command {
    "point": Point {
      "x": 0,
      "y": 0,
    },
    "type": "M",
  },
  Command {
    "cl": Point {
      "x": 0.2,
      "y": 0.2,
    },
    "cr": Point {
      "x": 0.6,
      "y": 0.8,
    },
    "point": Point {
      "x": 1,
      "y": 1,
    },
    "type": "C",
  },
  Command {
    "cl": Point {
      "x": 1.4,
      "y": 1.2,
    },
    "cr": Point {
      "x": 1.6,
      "y": 1.2,
    },
    "point": Point {
      "x": 2,
      "y": 1,
    },
    "type": "C",
  },
  Command {
    "cl": Point {
      "x": 2.4,
      "y": 0.8,
    },
    "cr": Point {
      "x": 2.8,
      "y": 0.2,
    },
    "point": Point {
      "x": 3,
      "y": 0,
    },
    "type": "C",
  },
]
`;

exports[`svg.ts Path commands parameter and getCommandString curve Normal 2`] = `"M 0 0 C 0.2 0.2 0.6 0.8 1 1 C 1.4 1.2 1.6 1.2 2 1 C 2.4 0.8 2.8 0.2 3 0"`;

exports[`svg.ts Path parseCommandString close command 1`] = `
Array [
  Command {
    "point": Point {
      "x": 0,
      "y": 0,
    },
    "type": "M",
  },
  Command {
    "point": Point {
      "x": 1,
      "y": 1,
    },
    "type": "L",
  },
  Command {
    "cl": Point {
      "x": 2,
      "y": 2,
    },
    "cr": Point {
      "x": 2,
      "y": 4,
    },
    "point": Point {
      "x": 6,
      "y": 0,
    },
    "type": "C",
  },
]
`;

exports[`svg.ts Path parseCommandString failed pattern 1`] = `Array []`;

exports[`svg.ts Path parseCommandString failed pattern 2`] = `
Array [
  Command {
    "point": Point {
      "x": 0,
      "y": 0,
    },
    "type": "M",
  },
]
`;

exports[`svg.ts Path parseCommandString success pattern 1`] = `
Array [
  Command {
    "point": Point {
      "x": 0,
      "y": 0,
    },
    "type": "M",
  },
  Command {
    "point": Point {
      "x": 1,
      "y": 1,
    },
    "type": "L",
  },
  Command {
    "cl": Point {
      "x": 2,
      "y": 2,
    },
    "cr": Point {
      "x": 2,
      "y": 4,
    },
    "point": Point {
      "x": 6,
      "y": 0,
    },
    "type": "C",
  },
]
`;

exports[`svg.ts Path toJson and toElement toElement 1`] = `
<path
  d="M 0 0 C 0.2 0.2 0.6 0.8 1 1 C 1.4 1.2 1.6 1.2 2 1 C 2.4 0.8 2.8 0.2 3 0"
  fill="none"
  stroke="#000"
  stroke-linecap="round"
  stroke-linejoin="round"
  stroke-width="1"
/>
`;

exports[`svg.ts Path toJson and toElement toJson 1`] = `
Object {
  "d": "M 0 0 C 0.2 0.2 0.6 0.8 1 1 C 1.4 1.2 1.6 1.2 2 1 C 2.4 0.8 2.8 0.2 3 0",
  "fill": "none",
  "stroke": "#000",
  "strokeLinecap": "round",
  "strokeLinejoin": "round",
  "strokeWidth": 1,
}
`;

exports[`svg.ts Svg parseSVGElement 1`] = `
Svg {
  "_paths": Array [
    Path {
      "close": true,
      "commands": Array [
        Command {
          "cl": undefined,
          "cr": undefined,
          "point": Point {
            "x": 2,
            "y": 2,
          },
          "type": "M",
        },
        Command {
          "cl": undefined,
          "cr": undefined,
          "point": Point {
            "x": 4,
            "y": 4,
          },
          "type": "L",
        },
        Command {
          "cl": Point {
            "x": 6,
            "y": 6,
          },
          "cr": Point {
            "x": 10,
            "y": 6,
          },
          "point": Point {
            "x": 14,
            "y": 6,
          },
          "type": "C",
        },
      ],
      "curve": true,
      "fill": "#f00",
      "smoothRatio": 0.2,
      "stroke": "#00f",
      "strokeWidth": 8,
    },
  ],
  "height": 400,
  "width": 400,
}
`;

exports[`svg.ts Svg parseSVGString 1`] = `
Svg {
  "_paths": Array [
    Path {
      "close": true,
      "commands": Array [
        Command {
          "cl": undefined,
          "cr": undefined,
          "point": Point {
            "x": 2,
            "y": 2,
          },
          "type": "M",
        },
        Command {
          "cl": undefined,
          "cr": undefined,
          "point": Point {
            "x": 4,
            "y": 4,
          },
          "type": "L",
        },
        Command {
          "cl": Point {
            "x": 6,
            "y": 6,
          },
          "cr": Point {
            "x": 10,
            "y": 6,
          },
          "point": Point {
            "x": 14,
            "y": 6,
          },
          "type": "C",
        },
      ],
      "curve": true,
      "fill": "#f00",
      "smoothRatio": 0.2,
      "stroke": "#00f",
      "strokeWidth": 8,
    },
  ],
  "height": 400,
  "width": 400,
}
`;

exports[`svg.ts Svg toBase64 1`] = `"data:image/svg+xml;base64,PHN2ZyB3aWR0aD0iNTAwIiBoZWlnaHQ9IjUwMCIgdmVyc2lvbj0iMS4xIiB4bWxucz0iaHR0cDovL3d3dy53My5vcmcvMjAwMC9zdmciPjxwYXRoIHN0cm9rZS13aWR0aD0iMSIgc3Ryb2tlPSIjMDAwIiBmaWxsPSJub25lIiBzdHJva2UtbGluZWpvaW49InJvdW5kIiBzdHJva2UtbGluZWNhcD0icm91bmQiIGQ9Ik0gMCAwIEMgMC4yIDAuMiAwLjYgMC44IDEgMSBDIDEuNCAxLjIgMS42IDEuMiAyIDEgQyAyLjQgMC44IDIuOCAwLjIgMyAwIj48L3BhdGg+PHBhdGggc3Ryb2tlLXdpZHRoPSIxIiBzdHJva2U9IiMwMDAiIGZpbGw9Im5vbmUiIHN0cm9rZS1saW5lam9pbj0ibWl0dGVyIiBzdHJva2UtbGluZWNhcD0ic3F1YXJlIiBkPSJNIDQgNCBMIDkgNCBMIDkgOCBMIDMgMCBaIj48L3BhdGg+PC9zdmc+"`;

exports[`svg.ts Svg toElement 1`] = `
<svg
  height="500"
  width="500"
>
  <path
    d="M 0 0 C 0.2 0.2 0.6 0.8 1 1 C 1.4 1.2 1.6 1.2 2 1 C 2.4 0.8 2.8 0.2 3 0"
    fill="none"
    stroke="#000"
    stroke-linecap="round"
    stroke-linejoin="round"
    stroke-width="1"
  />
  <path
    d="M 4 4 L 9 4 L 9 8 L 3 0 Z"
    fill="none"
    stroke="#000"
    stroke-linecap="square"
    stroke-linejoin="mitter"
    stroke-width="1"
  />
</svg>
`;

exports[`svg.ts Svg toJson 1`] = `
Object {
  "height": 500,
  "paths": Array [
    Object {
      "d": "M 0 0 C 0.2 0.2 0.6 0.8 1 1 C 1.4 1.2 1.6 1.2 2 1 C 2.4 0.8 2.8 0.2 3 0",
      "fill": "none",
      "stroke": "#000",
      "strokeLinecap": "round",
      "strokeLinejoin": "round",
      "strokeWidth": 1,
    },
    Object {
      "d": "M 4 4 L 9 4 L 9 8 L 3 0 Z",
      "fill": "none",
      "stroke": "#000",
      "strokeLinecap": "square",
      "strokeLinejoin": "mitter",
      "strokeWidth": 1,
    },
  ],
  "width": 500,
}
`;
